src/lang/builtins_source.ml

```diff
@@ -96,3 +96,43 @@ let () =
          Clock.force_init (fun x -> List.exists (fun y -> Oo.id x = Oo.id y) l)
        in
          Lang.list ~t:s_t (List.map (fun x -> Lang.source (x:>Source.source)) l))
+
+let () =
+  let kind = Lang.univ_t 1 in
+  add_builtin "source.dump"  ~cat:Liq
+    ~descr:"Immediately encode the whole contents of a source into a file."
+    ~flags:[Lang.Experimental]
+    [
+      "", Lang.format_t kind, None, Some "Encoding format." ;
+      "", Lang.string_t, None, Some "Name of the file.";
+      "", Lang.source_t (Lang.univ_t 1), None, Some "Source to encode";
+    ]
+    Lang.unit_t
+    (fun p ->
+       let format_val = Lang.assoc "" 1 p in
+       let format = Lang.to_format format_val in
+       let encoder =
+         try
+           Encoder.get_factory format
+         with
+         | Not_found -> raise (Lang_errors.Invalid_value (format_val, "Unsupported format"))
+       in
+       let kind = Encoder.kind_of_format format in
+       let file = Lang.to_string (Lang.assoc "" 2 p) in
+       let s = Lang.to_source (Lang.assoc "" 3 p) in
+       let encoder = encoder "source.dump" Meta_format.empty_metadata in
+       let frame = Frame.create kind in
+       let oc = open_out file in
+       Strings.iter (output_substring oc) encoder.Encoder.header;
+       s#get_ready [];
+       while s#is_ready do
```

Comment:
Have you thought about using track markers instead? It might be a more natural way of doing it, avoiding the use of `once` every time?

Reply:
Not really sure about how to get track markers, any hint? :)